src/core/util.ts

```diff
@@ -653,20 +653,131 @@ export function isPrimitive(obj: any): boolean {
+        return existed;
+    }
+    get(key: KEY): T | undefined {
+        return this.data[key];
+    }
+    set(key: KEY, value: T): this {
+        this.data[key] = value;
+        return this;
+    }
+    has(key: KEY): boolean {
+        return this.data.hasOwnProperty(key);
+    }
+    keys(): Iterator<KEY, undefined> {
+        const keys = Object.keys(this.data) as KEY[];
+        let i = 0;
+        let end = keys.length;
+        return {
+            next() {
+                if (i < end) {
+                    return {value: keys[i] as KEY, done: false};
+                }
+                return {done: true, value: undefined};
+            }
+        };
+    }
+    values(): Iterator<T, undefined> {
+        const data = this.data;
+        const keys = Object.keys(this.data) as KEY[];
+        let i = 0;
+        let end = keys.length;
+        return {
+            next() {
+                if (i < end) {
+                    return {value: data[keys[i]] as T, done: false};
+                }
+                return {done: true, value: undefined};
+            }
+        };
+    }
+    entries(): Iterator<[KEY, T], undefined> {
+        const data = this.data;
+        const keys = Object.keys(this.data) as KEY[];
+        let i = 0;
+        let end = keys.length;
+        return {
+            next() {
+                if (i < end) {
+                    return {value: [keys[i], data[keys[i]]] as [KEY, T], done: false};
+                }
+                return {done: true, value: undefined};
+            }
+        };
+    }
+    forEach(callback: (key: KEY, value: T) => void): void {
+        // This is a potential performance bottleneck, see details in
+        // https://github.com/ecomfe/zrender/issues/965, however it is now
+        // less likely to occur as we default to native maps when possible.
+        for (const key in this.data) {
+            if (this.data.hasOwnProperty(key)) {
+                callback(key, this.data[key]);
+            }
+        }
+    }
+
+    [Symbol.iterator](): Iterator<[KEY, T], undefined> {
+        return this.entries();
+    }
+}
+
+// We want to use native Map if it is available, but we do not want to polyfill the global scope
+// in case users ship their own polyfills or patch the native map object in any way.
+const isNativeMapSupported = typeof Map === 'function';
+function maybeNativeMap<T, KEY extends string | number = string | number>(): MapInterface<T, KEY> {
+    // Map may be a native class if we are running in an ES6 compatible environment.
+    // eslint-disable-next-line
+    return (isNativeMapSupported ? Map : MapPolyfill) as unknown as MapInterface<T, KEY>;
+}
 
 /**
  * @constructor
- * @param {Object} obj Only apply `ownProperty`.
+ * @param {Object} obj
  */
 export class HashMap<T, KEY extends string | number = string | number> {
-
-    data: {[key in KEY]: T} = {} as {[key in KEY]: T};
+    data: MapInterface<T, KEY> = maybeNativeMap<T, KEY>();
```

Comment:
`= maybeNativeMap<T, KEY>()` can be removed as it will be assigned in the constructor.